### PR TITLE
feat: update root AST on soft failure in `LanguageServer`

### PR DESCRIPTION
**Pull request by jaschdoc**
Depends on https://github.com/flix/flix/pull/5202

Related to https://github.com/flix/flix/issues/5026

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala [hunk 1]
```diff
@@ -338,15 +338,40 @@ class LanguageServer(port: Int, o: Options) extends WebSocketServer(new InetSock
             ("id" -> requestId) ~ ("status" -> "failure") ~ ("time" -> e) ~ ("result" -> results.map(_.toJSON))
           }
 
-        case failure =>
-          // Case 2: Compilation failed. Send back the error messages.
+        case SoftFailure(root, errors) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:341]: Gasp! Code duplication.. Good Henrik is not around.

What we use an or pattern `Success(...) | SoftFailure(...)` would that work?

Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:341]: Haha yes! 🙈 

This was just a sketch to see if it was even the correct thing to do. I'll refactor it next year! 😄

Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:341]: But in all seriousness I'll find a prettier solution :)

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:341]: I think and OR pattern could work.

Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:341]: I think an OR pattern has ambiguous variables that would be bound (`errors` could refer to both patterns) but I'll come up with some other way :)

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:341]: I thought that was the point. Does Scala not allow it?

Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:341]: Nope, that's also why we also just use `failure` as the error pattern

**Comment by magnus-madsen** [comment]
> What do you think about this solution? The Or-pattern doesn't work because it binds `root` in both
> 
> ```scala
> case Success(root) | SoftFailure(root, errors) => ???
> ```
> 
> and as mentioned, Scala doesn't allow that type of shadowing / punning.

That's disappointing.

I think what you did is fine.

**Comment by magnus-madsen** [comment]
Excited to see when we are able to survive a single error and still have VSCode support :)

**Pull request merged**